### ISSUE TITLE
Enhancement Request: Add First-Class Type Support in GDScript

### ISSUE THREAD

**Issue by DevSavvySerge** [godotengine/godot-proposals#10115]
### Describe the project you are working on

I'm working on a Vampire Survivor-like game where I intend to use best practices of Object-Oriented (OO) programming. This includes principles like polymorphism and dependency injection to create a dynamic and maintainable codebase.

### Describe the problem or limitation you are having in your project

Describe the problem or limitation you are having in your project

GDScript currently lacks first-class support for types. For instance, there is no Type type, and while there is an enumeration for built-in types, it is insufficient for more complex needs. This limitation prevents using types as dictionary keys, among other issues. As a workaround, I have to manually define type-related constants in each class, such as:

```gdscript

class_name Player
extends CharacterBody2D
const nameOf: StringName = StringName("Player")
```

This is cumbersome and makes it challenging to create dynamic code handling.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I propose adding first-class type support to GDScript. This would include introducing a Type type and enhancing the existing type system to allow types to be used as dictionary keys and other similar use cases. This feature would streamline the development process by reducing the need for manual workarounds and enabling more dynamic and flexible code.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Introduce a Type type in GDScript:

```gdscript

var my_type: Type = typeof(Player)
```

Allow types to be used as dictionary keys:

```gdscript

var type_dict: Dictionary = {}
type_dict[typeof(Player)] = "This is a player type"
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

The current workaround involves manually defining type-related constants in each class, which is neither efficient nor scalable. While possible, this approach is far from ideal and can lead to code that is harder to maintain and extend.

### Is there a reason why this should be core and not an add-on in the asset library?

First-class type support is a fundamental feature that would benefit many developers using GDScript. Implementing it as a core feature would ensure consistency, better integration with the language's existing type system, and improved performance compared to an add-on. It would also foster best practices in OO programming within the Godot engine.

**Comment by InkONat** [comment]
this would probably be a change only for a future Godot 5, it is a very big change

**Comment by Maximilian-Seitz** [comment]
**TL/DR**: Exposing `GDScriptNativeClass` to GDScript, along with its `get_name` function would allow a seamless implementation of a Type class, but even without this, it's still very much possible (except for a lack of auto-completion with native types).

-----------------

I've done some fiddling around, and found a relatively simple way of implementing this yourself in GDScript. My implementation should work perfectly fine for most use-cases, and the API could be extended fairly easily (by using the functionality offered by ClassDB, Script-class and functions accepting Variant.Type).

The API currently allows checking if a variant is an instance of a Type (/ "inherits" it), two Type instances can be compared (using `==`), and Types can be instantiated (as long as that makes sense for the Type). Adding the name of a type would be trivial, if it weren't for Script-based types not always having one. A Type can be received either through an instance of the Type (using `Type.of(instance)`), or some identifier (using `Type.from(identifier)`).

The `Type.from(identifier)` function takes either a `Variant.Type` (e.g. `Type.from(TYPE_INT)`), a `Script` (e.g. `Type.from(MyClass)`), or native-class name (e.g. `Type.from("Node")`). It is also *technically* possible to use a `GDScriptNativeClass` (e.g. `Type.from(Node)`), but since the API thereof is very restrictive, this has to create an instance from which to get the class-name, which is a pretty intense operation, and may even lead to side-effects. I've only included the last part here, since it shows that `GDScriptNativeClass` could benefit from being exposed to GDScript (along with its `get_name` function).

-----------------

Anyway, here's some tests for my implementation:

```GDScript
extends Node

func _ready() -> void:
	var int_type := Type.from(TYPE_INT)
	var node_type := Type.from("Node") # better than: Type.from(Node)
	var a_type := Type.from(A)
	var b_type := Type.from(B)
	
	var int_type_alt := Type.of(1)
	var node_type_alt := Type.of(Node.new())
	var a_type_alt := Type.of(A.new())
	var b_type_alt := Type.of(B.new())
	
	assert(int_type.is_instantiated_by(1))
	assert(node_type.is_instantiated_by(Area2D.new()))
	assert(a_type.is_instantiated_by(A.new()))
	assert(b_type.is_instantiated_by(B.new()))
	
	# This shouldn't be used, as it is expensive (and may have side-effects)
	assert(node_type == Type.from(Node))
	
	assert(int_type == int_type_alt)
	assert(node_type == node_type_alt)
	assert(a_type == a_type_alt)
	assert(b_type == b_type_alt)
	
	assert(a_type != b_type)
	
	assert(!a_type.is_instantiated_by(B.new()))
	assert(!b_type.is_instantiated_by(A.new()))
	
	print("Done!")


class A:
	pass

class B:
	pass
```

-----------------

And here's the implementation itself:
(If you have a use for it, feel free to use it under `CC0`)

```GDScript
@abstract class_name Type extends RefCounted


@abstract func is_instantiated_by(value: Variant) -> bool

@abstract func create_instance() -> Variant


static func of(value: Variant) -> Type:
	var variant_type := typeof(value)
	
	if variant_type == TYPE_OBJECT:
		var object := value as Object
		var script := object.get_script() as Script
		
		if script:
			return from_script(script)
		else:
			return from_native_class_name(object.get_class())
	else:
		return from_variant_type(variant_type)


static func from(type: Variant) -> Type:
	if type is Script:
		return from_script(type)
	elif type is int and type >= 0 and type < TYPE_MAX:
		return from_variant_type(type)
	elif type is String:
		return from_native_class_name(type)
	elif type is Object:
		## type must be a GDScriptNativeClass instance.
		## This should be avoided, as it creates an instance of the type.
		return from_native_class(type)
	
	printerr("Unknown input for Type.from()")
	
	return null

static func from_script(script: Script) -> Type:
	return _ScriptType.get_for(script)

static func from_variant_type(variant_type: Variant.Type) -> Type:
	return _VariantType.get_for(variant_type)

static func from_native_class_name(native_class_name: String) -> Type:
	return _NativeObjectType.get_for(native_class_name)

## WARNING: This should be avoided,
## since it instantiates the given class in order to get the name from it.
## There is currently no other way of getting this name from a GDScriptNativeClass.
static func from_native_class(native_class: Object) -> Type:
	if not native_class.get_class() == "GDScriptNativeClass":
		printerr("Unknown input for Type.from_native_class()")
		return null
	
	var instance: Object = native_class.new()
	var name := instance.get_class()
	instance.free()
	
	return from_native_class_name(name)


class _ScriptType extends Type:
	static var _script_types: Dictionary[Script, _ScriptType]
	
	var native_type: Script
	
	func _init(script: Script) -> void:
		native_type = script
		_script_types[script] = self
	
	static func get_for(script: Script) -> _ScriptType:
		if _script_types.has(script):
			return _script_types[script]
		else:
			return _ScriptType.new(script)
	
	func is_instantiated_by(value: Variant) -> bool:
		return is_instance_of(value, native_type)
	
	func create_instance() -> Variant:
		if not native_type.can_instantiate():
			return null
		var instance := Object.new()
		instance.set_script(native_type)
		return instance


class _NativeObjectType extends Type:
	static var _native_object_types: Dictionary[String, _NativeObjectType]
	
	var native_type: String
	
	func _init(type: String) -> void:
		native_type = type
		_native_object_types[type] = self
	
	static func get_for(type: String) -> _NativeObjectType:
		if not type_exists(type):
			return null
		
		if _native_object_types.has(type):
			return _native_object_types[type]
		else:
			return _NativeObjectType.new(type)
	
	func is_instantiated_by(value: Variant) -> bool:
		return (value is Object) and (value as Object).is_class(native_type)
	
	func create_instance() -> Variant:
		if not ClassDB.can_instantiate(native_type):
			return null
		return ClassDB.instantiate(native_type)


class _VariantType extends Type:
	static var _variant_types: Dictionary[Variant.Type, _VariantType]
	
	var native_type: Variant.Type
	
	func _init(type: Variant.Type) -> void:
		native_type = type
		_variant_types[type] = self
	
	static func get_for(type: Variant.Type) -> _VariantType:
		if _variant_types.has(type):
			return _variant_types[type]
		else:
			return _VariantType.new(type)
	
	func is_instantiated_by(value: Variant) -> bool:
		return is_instance_of(value, native_type)
	
	func create_instance() -> Variant:
		return type_convert(null, native_type)
```
